File: R/distn.R
## ID: distn.R, last updated 2023-12-30, F.Osorio and T.Wolodzko

dlaplace <- function(x, location = 0, scale = 1, log = FALSE)
{ # density of the Laplace distribution
  if (scale <= 0.0)
    stop("'scale' must be non-negative.")

  n <- length(x)
  nloc <- length(location)
  nscale <- length(scale)

  y <- .C("d_laplace",
          n = as.integer(n),
          y = double(n),
          x = as.double(x),
          location = as.double(location),
          nloc = as.integer(nloc),
          scale = as.double(scale),
          nscale = as.integer(nscale),
          give.log = as.integer(log))$y
  y
}

plaplace <- function(q, location = 0, scale = 1, lower.tail = TRUE, log.p = FALSE)
{ # distribution function of the Laplace distribution
  if (scale <= 0.0)
    stop("'scale' must be non-negative.")

  n <- length(q)
  nloc <- length(location)
  nscale <- length(scale)

  y <- .C("p_laplace",
          n = as.integer(n),
          y = double(n),
          q = as.double(q),
          location = as.double(location),
          nloc = as.integer(nloc),
          scale = as.double(scale),
          nscale = as.integer(nscale),
          lower.tail = as.integer(lower.tail),
          log.p = as.integer(log.p))$y
  y
}

qlaplace <- function(p, location = 0, scale = 1, lower.tail = TRUE, log.p = FALSE)
{ # quantile function of the Laplace distribution
  if (scale <= 0.0)
    stop("'scale' must be non-negative.")

  n <- length(p)
  nloc <- length(location)
  nscale <- length(scale)

  y <- .C("q_laplace",
          n = as.integer(n),
          y = double(n),
          p = as.double(p),
          location = as.double(location),
          nloc = as.integer(nloc),
          scale = as.double(scale),
          nscale = as.integer(nscale),
          lower.tail = as.integer(lower.tail),
          log.p = as.integer(log.p))$y
  y
}

dmLaplace <-
function(x, center = rep(0, nrow(Scatter)), Scatter = diag(length(center)), log = FALSE)
{ # pdf for the multivariate Laplace distribution
  give.log <- log
  if (is.data.frame(x))
    x <- as.matrix(x)
  else if (!is.matrix(x))
    stop("'x' must be a matrix or a data frame")
  if (!all(is.finite(x)))
    stop("'x' must contain finite values only")
  n <- nrow(x)
  p <- ncol(x)

  if (is.null(center))
    stop("'center' must be provided")
  if (isFALSE(center))
    center <- double(p) # center is zeroed
  if (!is.vector(center))
    stop("'center' must be a vector")
  if (length(center) != p)
    stop("'center' has incorrect length")

  if (is.null(Scatter))
    stop("'Scatter' matrix must be provided")
  if (!is.matrix(Scatter))
    stop("'Scatter' must be a matrix")
  if (all(dim(Scatter) != c(p,p)))
    stop("'Scatter' has incorrect dimensions")
  if (!isSymmetric(Scatter))
    Scatter <- asSymmetric(Scatter)

  storage.mode(x) <- "double"
  storage.mode(Scatter) <- "double"

  y <- .C("pdf_mlaplace",
          y = double(n),
          x = x,
          n = as.integer(n),
          p = as.integer(p),
          center = as.double(center),
          Scatter = Scatter)$y

  if (!give.log) y <- exp(y)
  y
}

File: src/distn.c
/* ID: distn.c, last updated 2023-12-28, F.Osorio */

#include "base.h"
#include "interface.h"
#include "lad.h"

/* dpq-functions for Laplace distribution */
static double pdf_laplace(double, double, double, int);
static double cdf_laplace(double, double, double, int, int);
static double quantile_laplace(double, double, double, int, int);
/* ..end declarations */

/* ========================================================================== *
 * pdf, cdf and quantile functions for the univariate Laplace distribution
 * ========================================================================== */

static double pdf_laplace(double x, double location, double scale, int log_pdf)
{ /* density of the Laplace distribution */
  double y;

  y = fabs(x - location) / scale;

  return (log_pdf ?
    -.5 * M_LN2 - log(scale) - M_SQRT2 * y :
    M_SQRT1_2 * exp(-M_SQRT2 * y) / scale);
}

static double cdf_laplace(double x, double location, double scale, int lower, int log_cdf)
{ /* distribution function of the Laplace distribution */
  double val;

  x = (x - location);

  if (x < 0.)
    val = .5 * exp(M_SQRT2 * x / scale);
  else 
    val = 1. - .5 * exp(-M_SQRT2 * x / scale);

  if (log_cdf) {
    return log(lower ? val : 1. - val);
  } else {
    return (lower ? val : 1. - val);
  }
}

static double quantile_laplace(double p, double location, double scale, int lower, int log_prob)
{ /* quantile function of the Laplace distribution */
  double q, val;

  if (scale == 0.)
    return location;

  if (log_prob)
    p = exp(p);

  if (!lower)
    p = 1. - p;

  q = p;

  if (q == .5)
    return location;

  if (q < .5)
    val = location + scale * M_SQRT1_2 * log(2. * q);
  else
    val = location - scale * M_SQRT1_2 * log(2. * (1. - q));

  return val;
}

void d_laplace(int *n, double *y, double *x, double *location, int *nloc, double *scale, int *nscale, int *give_log)
{ /* pdf of univariate Laplace distribution */
  int nobs = *n, na = *nloc, nb = *nscale, log_pdf = *give_log;

  for (int i = 0; i < nobs; i++)
    y[i] = pdf_laplace(x[i], location[i % na], scale[i % nb], log_pdf);
}

void p_laplace(int *n, double *y, double *x, double *location, int *nloc, double *scale, int *nscale, int *lower_tail, int *log_p)
{ /* cdf of univariate Laplace distribution */
  int nobs = *n, na = *nloc, nb = *nscale, lower = *lower_tail, log_cdf = *log_p;

  for (int i = 0; i < nobs; i++)
    y[i] = cdf_laplace(x[i], location[i % na], scale[i % nb], lower, log_cdf);
}

void q_laplace(int *n, double *y, double *p, double *location, int *nloc, double *scale, int *nscale, int *lower_tail, int *log_p)
{ /* quantile function of univariate Laplace distribution */
  int nobs = *n, na = *nloc, nb = *nscale, lower = *lower_tail, log_prob = *log_p;

  for (int i = 0; i < nobs; i++)
    y[i] = quantile_laplace(p[i], location[i % na], scale[i % nb], lower, log_prob);
}

/* ========================================================================== *
 * pdf for the multivariate Laplace distribution
 * ========================================================================== */

void
pdf_mlaplace(double *y, double *x, int *nobs, int *vars, double *center, double *Scatter)
{ /* evaluation of the multivariate Laplace log-density function */
  int errcode = 0, job = 0, n = *nobs, p = *vars;
  double *Root, *z, D2, log_pdf;

  Root = (double *) Calloc(p * p, double);
  z    = (double *) Calloc(p, double);

  copy_lower(Root, p, Scatter, p, p);
  chol_decomp(Root, p, p, job, &errcode);
  if (errcode)
    error("Cholesky decomposition in pdf_mlaplace gave code %d", errcode);
  
  log_pdf  = lgammafn(0.5 * p) - (double) p * M_LN_SQRT_PI - lgammafn((double) p) - (p + 1.0) * M_LN2;
  log_pdf -= logAbsDet(Root, p, p);
  
  for (int i = 0; i < n; i++) {
    copy_vec(z, 1, x + i, n, p);
    D2 = mahalanobis(z, p, center, Root);
    y[i] = log_pdf - 0.5 * sqrt(D2);
  }

  Free(Root); Free(z);
}
